File: dsa/concepts/04_twopointers/java/FindtheUnionOfTwoSortedArrays.java

import java.util.ArrayList;
import java.util.List;

public class FindtheUnionOfTwoSortedArrays {

    public static void main(String[] args) {
        FindtheUnionOfTwoSortedArrays obj = new FindtheUnionOfTwoSortedArrays();
        int[] nums1 = { 1, 2, 3, 4, 5 };
        int[] nums2 = { 1, 2, 3 };
        System.out.println(obj.findUnion(nums1, nums2));
    }

    // Problem: Find the union of two sorted arrays
    public List<Integer> findUnion(int[] nums1, int[] nums2){
        int n1 = nums1.length, n2 = nums2.length;
        int i=0, j=0;
        List<Integer> union = new ArrayList<>();
        while(i<n1 && j<n2){
            if(nums1[i] <= nums2[j]){
                if(union.isEmpty() || union.get(union.size()-1) != nums1[i]){
                    union.add(nums1[i]);
                }
                i++;
            }else{
                if(union.isEmpty() || union.get(union.size()-1) != nums2[j]){
                    union.add(nums2[j]);
                }
                j++;
            }
        }
        while(i<n1){
            if(union.isEmpty() || union.get(union.size()-1) != nums1[i]){
                union.add(nums1[i]);
            }
            i++;
        }
        while(j<n2){
            if(union.isEmpty() || union.get(union.size()-1) != nums2[j]){
                union.add(nums2[j]);
            }
            j++;
        }
        return union;
    }
}

File: dsa/concepts/04_twopointers/python/find_the_union_of_two_sorted_arrays.py
"""
Find the Union of Two Sorted Arrays
Problem: Find the union of two sorted arrays
Idea: Use two pointers to traverse both arrays and avoid duplicates
Time: O(n + m) where n and m are lengths of the arrays
Space: O(1) excluding the result array
"""

def find_union(nums1, nums2):
    """
    Find the union of two sorted arrays.
    
    Args:
        nums1: List[int] - First sorted array
        nums2: List[int] - Second sorted array
    
    Returns:
        List[int] - List containing union elements without duplicates
    """
    n1, n2 = len(nums1), len(nums2)
    i = j = 0
    union = []
    
    while i < n1 and j < n2:
        if nums1[i] <= nums2[j]:
            if not union or union[-1] != nums1[i]:
                union.append(nums1[i])
            i += 1
        else:
            if not union or union[-1] != nums2[j]:
                union.append(nums2[j])
            j += 1
    
    while i < n1:
        if not union or union[-1] != nums1[i]:
            union.append(nums1[i])
        i += 1
    
    while j < n2:
        if not union or union[-1] != nums2[j]:
            union.append(nums2[j])
        j += 1
    
    return union


if __name__ == "__main__":
    nums1 = [1, 2, 3, 4, 5]
    nums2 = [1, 2, 3]
    result = find_union(nums1, nums2)
    print(result)  # Expected: [1, 2, 3, 4, 5]
